internal/distributed/connection_manager.go

```diff
@@ -0,0 +1,309 @@
+
+	return nil
+}
+
+func (cm *ConnectionManager) GetClients(serverName string) (interface{}, bool) {
+	if !cm.checkServerName(serverName) {
+		return nil, false
+	}
+
+	_, ok := cm.dependencies[serverName]
+	if !ok {
+		log.Error("Dependency has not been added yet", zap.Any("serverName", serverName))
+		return nil, false
+	}
+
+	switch serverName {
+	case typeutil.RootCoordRole:
+		return cm.rootCoord, true
+	case typeutil.DataCoordRole:
+		return cm.dataCoord, true
+	case typeutil.IndexCoordRole:
+		return cm.indexCoord, true
+	case typeutil.QueryCoordRole:
+		return cm.queryCoord, true
+	case typeutil.DataNodeRole:
+		return cm.dataNodes, true
+	case typeutil.IndexNodeRole:
+		return cm.indexNodes, true
+	case typeutil.QueryNodeRole:
+		return cm.queryNodes, true
+	}
+	return nil, false
+}
+
+func (cm *ConnectionManager) Stop() {
+	for _, task := range cm.buildTasks {
+		task.Stop()
+	}
+	close(cm.closeCh)
+}
+
+func (cm *ConnectionManager) processEvent(channel <-chan *sessionutil.SessionEvent) {
+	for {
+		select {
+		case _, ok := <-cm.closeCh:
+			if !ok {
+				return
+			}
+		case ev := <-channel:
+			if ev.EventType == sessionutil.SessionAddEvent {
+				cm.buildConnections(ev.Session)
+			} else if ev.EventType == sessionutil.SessionDelEvent {
+				cm.removeTask(ev.Session.ServerID)
+			}
+		}
+	}
+}
+
+func (cm *ConnectionManager) receiveFinishTask() {
+	go func() {
+		for {
+			select {
+			case _, ok := <-cm.closeCh:
+				if !ok {
+					return
+				}
+			case serverID := <-cm.notify:
+				cm.taskMu.Lock()
+				task, ok := cm.buildTasks[serverID]
+				if ok {
+					if task.state == buildConnectionsuccess {
+						cm.buildClients(task.sess, task.result)
+					}
+					delete(cm.buildTasks, serverID)
+				}
+				cm.taskMu.Unlock()
+			}
+		}
+	}()
+}
+
+func (cm *ConnectionManager) buildClients(session *sessionutil.Session, connection *grpc.ClientConn) {
+	switch session.ServerName {
+	case typeutil.RootCoordRole:
+		//cm.rootCoord = grpcrootcoordclient.NewClient(connection)
+	case typeutil.DataCoordRole:
+		//cm.dataCoord = grpcdatacoordclient.NewClient(connection)
+	case typeutil.IndexCoordRole:
+		//cm.indexCoord = grpcindexcoord.NewClient(connection)
+	case typeutil.QueryCoordRole:
+		//cm.queryCoord = grpcquerycoord.NewClient(connection)
+	case typeutil.DataNodeRole:
+		//cm.dataNodes[session.ServerID] = grpcdatanodeclient.NewClient(connection)
+	case typeutil.IndexNodeRole:
+		//cm.indexNodes[session.ServerID] = grpcindexnodeclient.NewClient(connection)
+	case typeutil.QueryNodeRole:
+		//cm.queryNodes[session.ServerID] = grpcquerynodeclient.NewClient(connection)
+	}
+}
+
+func (cm *ConnectionManager) buildConnections(session *sessionutil.Session) {
+	task := newBuildClientTask(session, cm.notify)
+	cm.buildTasks[session.ServerID] = task
+	task.Run()
+}
+
+func (cm *ConnectionManager) removeTask(id int64) {
+	cm.taskMu.Lock()
+	task, ok := cm.buildTasks[id]
+	if ok {
+		task.Stop()
+		delete(cm.buildTasks, id)
```

Comment:
should not only remove buildTasks, we should also has a cache of all grpc client, remove from the list